pkg/plugins/runtime/gateway/listener_generator.go

```diff
@@ -1,21 +1,243 @@
+		return nil
+	}
+
+	candidates := make([]policy.DataplanePolicy, len(gatewayList.Items))
+	for i, gw := range gatewayList.Items {
+		candidates[i] = gatewayPolicyAdaptor{gw}
+	}
+
+	if p := policy.SelectDataplanePolicy(dp, candidates); p != nil {
+		return p.(gatewayPolicyAdaptor).GatewayResource
+	}
+
+	return nil
+}
+
+var _ policy.DataplanePolicy = gatewayPolicyAdaptor{}
+
+// ListenerGenerator generates Kuma gateway listeners.
 type ListenerGenerator struct {
 	Resources manager.ReadOnlyResourceManager
 }
 
 var _ generator.ResourceGenerator = ListenerGenerator{}
 
-func (ListenerGenerator) Generate(ctx xds_context.Context, proxy *xds.Proxy) (*xds.ResourceSet, error) {
-	return nil, errors.New("not implemented")
+func (l ListenerGenerator) Generate(_ xds_context.Context, proxy *core_xds.Proxy) (*core_xds.ResourceSet, error) {
+	gw := findMatchingGateway(l.Resources, proxy.Dataplane)
+	if gw == nil {
+		log.V(1).Info("no matching gateway for dataplane",
+			"name", proxy.Dataplane.Meta.GetName(),
+			"mesh", proxy.Dataplane.Meta.GetMesh(),
+		)
+
+		return nil, nil
+	}
+
+	log.V(1).Info(fmt.Sprintf("matched gateway %q to dataplane %q",
+		gw.Meta.GetName(), proxy.Dataplane.Meta.GetName()))
+
+	// Multiple listener specifications can have the same port. If
+	// they are compatible, then we can collapse those specifications
+	// down to a single listener.
+	collapsed := map[uint32][]*mesh_proto.Gateway_Listener{}
+	for _, ep := range gw.Spec.GetConf().GetListeners() {
+		collapsed[ep.GetPort()] = append(collapsed[ep.GetPort()], ep)
+	}
+
+	// A Gateway is a single service across all listeners.
+	service := proxy.Dataplane.Spec.GetIdentifyingService()
+
+	resources := core_xds.NewResourceSet()
+
+	// Generate a listener resource for each port.
+	for port, listeners := range collapsed {
+		protocol := listeners[0].GetProtocol()
+		address := proxy.Dataplane.Spec.GetNetworking().Address
+
+		log.V(1).Info("generating listener",
+			"address", address,
+			"port", port,
+			"protocol", protocol,
+		)
+
+		// TODO(jpeach) verify that the listeners are compatible.
+		// TODO(jpeach) hoist the compatibility check and use it in Gateway validation.
+
+		// This check forces all listeners on the port to have
+		// the same protocol, which is unnecessarily strict. We
+		// cal allow TLS and HTTPS on the same port, for example.
+		for i := range listeners {
+			if listeners[i].GetProtocol() != listeners[0].GetProtocol() {
+				return nil, errors.Errorf("cannot collapse listener protocols %s and %s",
+					listeners[i].GetProtocol(), listeners[0].GetProtocol(),
+				)
+			}
+		}
+
+		switch protocol {
+		case mesh_proto.Gateway_Listener_UDP:
+			fallthrough
+		case mesh_proto.Gateway_Listener_TCP:
+			fallthrough
+		case mesh_proto.Gateway_Listener_TLS:
+			fallthrough
+		case mesh_proto.Gateway_Listener_HTTPS:
+			return nil, errors.Errorf("unsupported protocol %q", protocol)
```

Comment:
nit: you can do `case mesh_proto.Gateway_Listener_UDP, mesh_proto.Gateway_Listener_TCP, mesh_proto.Gateway_Listener_TLS, mesh_proto.Gateway_Listener_HTTPS`

Reply:
Yeh, it's  like this because at some point each of those branches will have to do something different.